src/v2/providers/eventarc.ts

```diff
@@ -0,0 +1,163 @@
+  eventType: string,
+  handler: CloudEventHandler
+): CloudFunction<CloudEvent<T>>;
+
+export function onCustomEventPublished<T = any>(
+  eventType: string,
+  opts: EventarcTriggerOptions,
+  handler: CloudEventHandler
+): CloudFunction<CloudEvent<T>>;
+
+export function onCustomEventPublished<T = any>(
+  eventType: string,
+  optsOrHandler:
+    CloudEventHandler
+    | EventarcTriggerOptions,
+  handler?: CloudEventHandler,
+): CloudFunction<CloudEvent<T>> {
+  if (arguments.length < 2) {
+    throw new Error(
+      'Missing required parameters.'
+    );  
+  }
+  if (arguments.length == 2) {
+    if (typeof optsOrHandler !== 'function') { 
+      throw new Error(
+        'Expected second parameter to be the event handler function.'
+      );
+    }
+    handler = optsOrHandler as CloudEventHandler;
+    optsOrHandler = {} as EventarcTriggerOptions;
+  } else if (arguments.length == 3) {
+    optsOrHandler = optsOrHandler as EventarcTriggerOptions;
+  }
+
+  const opts = optsOrHandler as EventarcTriggerOptions;
+  const func = (raw: CloudEvent<unknown>) => {
+    return handler(raw as CloudEvent<T>);
+  };
+
+  func.run = handler;
+
+  const channel = opts.channel ?? 'locations/us-central1/channels/firebase';
+
+  Object.defineProperty(func, '__trigger', {
+    get: () => {
+      const baseOpts = options.optionsToTriggerAnnotations(
+        options.getGlobalOptions()
+      );
+      const specificOpts = options.optionsToTriggerAnnotations(opts);
+      const trigger = {
+        platform: 'gcfv2',
+        ...baseOpts,
+        ...specificOpts,
+        labels: {
+          ...baseOpts?.labels,
+          ...specificOpts?.labels,
+        },
+        eventTrigger: {
+          eventType,
+          filters: opts.filters,
+          channel,
+        },
+      };
+      if (!opts.filters) {
+        delete trigger.eventTrigger.filters;
+      }
+
+      return trigger;
+    },
+  });
+
+  // TypeScript doesn't recognize defineProperty as adding a property and complains
+  // that __endpoint doesn't exist. We can either cast to any and lose all type safety
+  // or we can just assign a meaningless value before calling defineProperty.
+  func.__endpoint = {} as ManifestEndpoint;
+
+  // SDK may attempt to read FIREBASE_CONFIG env var to fetch the default bucket name.
+  // To prevent runtime errors when FIREBASE_CONFIG env var is missing, we use getters.
+  Object.defineProperty(func, '__endpoint', {
+    get: () => {
+      const baseOpts = options.optionsToEndpoint(options.getGlobalOptions());
+      const specificOpts = options.optionsToEndpoint(opts);
+
+      const endpoint: ManifestEndpoint = {
+        platform: 'gcfv2',
+        ...baseOpts,
+        ...specificOpts,
+        labels: {
+          ...baseOpts?.labels,
+          ...specificOpts?.labels,
+        },
+        eventTrigger: {
+          eventType,
+          eventFilters: opts.filters,
+          retry: false,
+          channel,
```

Comment:
Per container contract, `channel` isn't allowed and we got invalid EventFilters.

I think we instead need this to be something like:

```
const eventFilters = [
  { attribute: "channel", value: channel }
  ...Object.entries(opts.filters).map((k, v) => { attribute: k, value: v })
]
```

Reply:
Channel is not a filter, it's a top level trigger field in GCP CP. I guess I should add this to the container contract?

Reply:
This is implemented in 

1. https://github.com/firebase/firebase-functions/pull/1070
2. https://github.com/firebase/firebase-tools/pull/4388

Reply:
Updated. Thanks.